### PR TITLE
Move ReadMe Docs to KDocs

### PR DESCRIPTION
**Pull request by maruiz81**
Data types included: 
* Option
* Either
* Try
* Validated
* NonEmptyLIst

### CODE REVIEW

**Comment by i-walker** [comment]
The FlowableK tests are sometimes flakey. Let's check if everything builds correctly and see what is missing

File: modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Bitraverse.kt [hunk 1]
```diff
@@ -39,7 +39,6 @@ import arrow.typeclasses.internal.IdBimonad
  * ## Bisequence
  *
  * `Bisequence` invert the original structure `F<G<A,B>>` to `G<F<A>,F<B>>`
- * `Bisequence` invert the original structure `F<G<A,B>>` to `G<F<A>,F<B>>`
```
Comment (i-walker) [inline modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Bitraverse.kt]: Bitraverse was from your PR right?

Comment (maruiz81) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Bitraverse.kt]: yes, I did last week, but I have realised I put this duplicate line

**Comment by maruiz81** [comment]
@i-walker have you checked my question here https://github.com/arrow-kt/arrow/issues/1490#issuecomment-504681807?

Thanks

**Comment by i-walker** [comment]
May you write to me on the kotlin slack - Imran S. [Arrow]. I don't really know what the exact problem is.

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/Option.kt [hunk 2]
```diff
@@ -3,9 +3,349 @@ package arrow.core
...
+ * ```kotlin:ank:silent
+ * fun maybeItWillReturnSomething(flag: Boolean): Option<String> =
+ *  if (flag) Some("Found value") else None
+ * ```
+ *
+ * Using `getOrElse` we can provide a default value `"No value"` when the optional argument `None` does not exist:
+ *
+ * ```kotlin:ank:playground
+ * import arrow.core.*
+ * fun maybeItWillReturnSomething(flag: Boolean): Option<String> =
```
[the remark this answers is not in the document]
Comment (maruiz81) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/core/Option.kt]: If I remove it from here when I execute in the website I get the error "Unresolved reference"
image

[the remark this answers is not in the document]
Comment (raulraja) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/core/Option.kt]: @i-walker declarations in previous snippets never work in playgrounds. Playgrounds need to include all imports and code they depend on.

[the remark this answers is not in the document]
Comment (i-walker) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/core/Option.kt]: 😅 sorry, guys haha. I actually forgot that.

Hi @maruiz81 then, please wrap the function after the first declaration outside of the //sampleStart .. //sampleEnd

[the remark this answers is not in the document]
Comment (maruiz81) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/core/Option.kt]: Hi @i-walker, I don't know if I got you, that is actually outside the block //sampleStart...//sampleEnd

[the remark this answers is not in the document]
Comment (i-walker) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/core/Option.kt]: I'll ping you in slack

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/Option.kt [hunk 3]
```diff
@@ -3,9 +3,349 @@ package arrow.core
...
+ * import arrow.core.*
+ * fun main() {
+ *  //sampleStart
+ *  val value: Option<String> = Some("I am wrapped in something")
+ *  //sampleEnd
+ *  println("value = $value")
+ * }
+ * ```
+ *
+ * ```kotlin:ank:playground
```
[the remark this answers is not in the document]
Comment (maruiz81) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/core/Option.kt]: done!

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt [hunk 4]
```diff
@@ -7,7 +7,203 @@ import arrow.typeclasses.Applicative
...
+ *
+ * A `NonEmptyList` guarantees the list always has at least 1 element.
+ *
+ * ```kotlin:ank:playground
+ * import arrow.core.*
+ *
+ * fun main() {
+ * val value =
+ * //sampleStart
+ * //NonEmptyList.of() // does not compile
```
Comment (i-walker) [inline modules/core/arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt]:
```suggestion
 * // NonEmptyList.of() // does not compile
```

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/Option.kt [hunk 5]
```diff
@@ -3,9 +3,340 @@ package arrow.core
...
+ * Using `getOrElse` we can provide a default value `"No value"` when the optional argument `None` does not exist:
+ *
+ * ```kotlin:ank:playground
+ * import arrow.core.*
+ * fun maybeItWillReturnSomething(flag: Boolean): Option<String> =
+ *  if (flag) Some("Found value") else None
+ *
+ *  fun main() {
+ * //sampleStart
+ *  val option = maybeItWillReturnSomething(true)
```
Comment (i-walker) [inline modules/core/arrow-core-data/src/main/kotlin/arrow/core/Option.kt]: lift the option value outside of the main block

**Comment by i-walker** [comment]
I'll catch up with your PR, after the changes we discussed in Slack

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/Option.kt [hunk 6]
```diff
@@ -3,9 +3,392 @@ package arrow.core
...
+ * fun main() {
+ *  println("value = $value")
+ * }
+ * ```
+ *
+ * ### Supported type classes
+ *
+ * ```kotlin:ank:replace
+ * import arrow.reflect.*
+ * import arrow.core.*
```
Comment (i-walker) [inline modules/core/arrow-core-data/src/main/kotlin/arrow/core/Option.kt]: please provide the specific imports

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/Validated.kt [hunk 7]
```diff
@@ -10,7 +10,590 @@ typealias Valid<A> = Validated.Valid<A>
...
+ *  println(houseNumber)
+ * }
+ *
+ * ```
+ *
+ * ## Alternative validation strategies to Validated: using `ApplicativeError`
+ *
+ * We may use `ApplicativeError` instead of `Validated` to abstract away validation strategies and raising errors in the context we are computing in.
+ *
+ * ```kotlin:ank
```
Comment (i-walker) [inline modules/core/arrow-core-data/src/main/kotlin/arrow/core/Validated.kt]: can you turn this into a playground snippet :)

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/Validated.kt [hunk 8]
```diff
@@ -10,7 +10,590 @@ typealias Valid<A> = Validated.Valid<A>
...
+ *   Rules<EitherPartialOf<Nel<ValidationError>>>(Either.applicativeError())
+ *
+ *  companion object {
+ *   infix fun <A> failFast(f: FailFastStrategy.() -> A): A = f(FailFastStrategy)
+ *   infix fun <A> accumulateErrors(f: ErrorAccumulationStrategy.() -> A): A = f(ErrorAccumulationStrategy)
+ *  }
+ * }
+ *
+ * //sampleStart
+ * val value =Rules failFast {
```
Comment (i-walker) [inline modules/core/arrow-core-data/src/main/kotlin/arrow/core/Validated.kt]:
```suggestion
 * val value = Rules failFast {
```

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/Validated.kt [hunk 9]
```diff
@@ -10,7 +10,590 @@ typealias Valid<A> = Validated.Valid<A>
...
+ * fun main() {
+ *  println("value = $value")
+ * }
+ * ```
+ *
+ * ### Supported type classes
+ *
+ * ```kotlin:ank:replace
+ * import arrow.reflect.*
+ * import arrow.core.*
```
Comment (i-walker) [inline modules/core/arrow-core-data/src/main/kotlin/arrow/core/Validated.kt]: add specific import

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/Option.kt [hunk 10]
```diff
@@ -3,9 +3,392 @@ package arrow.core
 import arrow.higherkind
 
 /**
- * Represents optional values. Instances of `Option`
- * are either an instance of $some or the object $none.
+ *
+ * ank_macro_hierarchy(arrow.core.Option)
+ *
+ * {:.beginner}
```
[the remark this answers is not in the document]
Comment (maruiz81) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/core/Option.kt]: Why has it to be deleted?

**Comment by maruiz81** [comment]
> please construct for your values something like this
> 
> ```
> val value: Int =
> //sampleStart
>     4
> //sampleEnd
> ```
> 
> so Users can't change the print value identifier. Any other intermediate variables can stay as there are.
> 
> Also we need coherence wether there is spacing between //sampleEnd and the main function or not.
> 
> Thank you buddy. Feel free to ask, if something seems odd.

I had the code in the way you have told me but I changed to be consistent because I have many places  where I have more than one variable, something like this:

```
//sampleStart
val valueA: Int = 4
val valueB: Int = 5
//sampleEnd
```
So basically do you prefer to have the declaration outside only in the cases you only have a variable? Or do you want to have all the snipped in the same way?

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/Option.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (i-walker) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/core/Option.kt]: don't delete this.
I checked in with other maintainers :)
I forgot to delete this review part

**Comment by i-walker** [comment]
This format is for every value which is later printed.
For instance both values need to be printed, then do this:
```
```kotlin:ank:playground
val valueA: Int =
//sampleStart
    4
//sampleEnd
val valueB: Int =
//sampleStart
    5
//sampleEnd
fun main(){
    println(valueA)
    prinln(valueB)
}
```
And as above if only one is printed:
```
```kotlin:ank:playground
val valueA: Int = 
//sampleStart
    4
//sampleEnd
val valueB: Int = 5
fun main(){
 println(valueA)
}
```

**Comment by maruiz81** [comment]
> This format is for every value which is later printed.
> For instance both values need to be printed, then do this:
> 
> ```
> ```kotlin:ank:playground
> val valueA: Int =
> //sampleStart
>     4
> //sampleEnd
> val valueB: Int =
> //sampleStart
>     5
> //sampleEnd
> fun main(){
>     println(valueA)
>     prinln(valueB)
> }
> ```
> 
> And as above if only one is printed:
> 
> ```
> ```kotlin:ank:playground
> val valueA: Int = 
> //sampleStart
>     4
> //sampleEnd
> val valueB: Int = 5
> fun main(){
>  println(valueA)
> }
> ```

I think I already checked to have more than one sample block and it didn't work, let me try again this afternoon.

**Comment by maruiz81** [comment]
> > This format is for every value which is later printed.
> > For instance both values need to be printed, then do this:
> > ```
> > ```kotlin:ank:playground
> > val valueA: Int =
> > //sampleStart
> >     4
> > //sampleEnd
> > val valueB: Int =
> > //sampleStart
> >     5
> > //sampleEnd
> > fun main(){
> >     println(valueA)
> >     prinln(valueB)
> > }
> > ```
> > 
> > 
> > And as above if only one is printed:
> > ```
> > ```kotlin:ank:playground
> > val valueA: Int = 
> > //sampleStart
> >     4
> > //sampleEnd
> > val valueB: Int = 5
> > fun main(){
> >  println(valueA)
> > }
> > ```
> 
> I think I already checked to have more than one sample block and it didn't work, let me try again this afternoon.

Yes the second block is ignored and the second `//sampleStart` and `//sampleEnd` are interpreted as comments.
image

**Comment by i-walker** [comment]
> > > This format is for every value which is later printed.
> > > For instance both values need to be printed, then do this:
> > > ```
> > > ```kotlin:ank:playground
> > > val valueA: Int =
> > > //sampleStart
> > >     4
> > > //sampleEnd
> > > val valueB: Int =
> > > //sampleStart
> > >     5
> > > //sampleEnd
> > > fun main(){
> > >     println(valueA)
> > >     prinln(valueB)
> > > }
> > > ```
> > > 
> > > 
> > > And as above if only one is printed:
> > > ```
> > > ```kotlin:ank:playground
> > > val valueA: Int = 
> > > //sampleStart
> > >     4
> > > //sampleEnd
> > > val valueB: Int = 5
> > > fun main(){
> > >  println(valueA)
> > > }
> > > ```
> > 
> > 
> > I think I already checked to have more than one sample block and it didn't work, let me try again this afternoon.
> 
> Yes the second block is ignored and the second `//sampleStart` and `//sampleEnd` are interpreted as comments.
> image

Ok keep the value identifieres as before. This won't work in the current impl. of ank.

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt [hunk 11]
```diff
@@ -7,7 +7,205 @@ import arrow.typeclasses.Applicative
...
+ * - We can easily construct values of `NonEmptyList` with `NonEmptyList.of`
+ * - `foldLeft`, `map`, `flatMap` and others are used to compute over the internal contents of a `NonEmptyList` value.
+ * - `fx { ... } comprehensions` can be __used to imperatively compute__ over multiple `NonEmptyList` values in sequence.
+ * - `NonEmptyList.applicative().map { ... }` can be used to compute over multiple `NonEmptyList` values preserving type information and __abstracting over arity__ with `map`
+ *
+ * ### Supported type classes
+ *
+ * ```kotlin:ank:replace
+ * import arrow.reflect.*
+ * import arrow.core.*
```
[the remark this answers is not in the document]
Comment (maruiz81) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt]: ok

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/Either.kt [hunk 12]
```diff
@@ -6,11 +6,621 @@ import arrow.core.Either.Right
...
 /**
- * Port of https://github.com/scala/scala/blob/v2.12.1/src/library/scala/util/Either.scala
+ * ---
+ *
+ * ank_macro_hierarchy(arrow.core.Either)
+ *
+ * {:.beginner}
+ * beginner
+ *
+ * [Перевод на русский](/docs/arrow/core/either/ru/)
```
Comment (i-walker) [inline modules/core/arrow-core-data/src/main/kotlin/arrow/core/Either.kt]: you can delete that reference 
```suggestion
```

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/Either.kt [hunk 13]
```diff
@@ -6,11 +6,620 @@ import arrow.core.Either.Right
...
 
 /**
- * Port of https://github.com/scala/scala/blob/v2.12.1/src/library/scala/util/Either.scala
+ * ---
+ *
+ * ank_macro_hierarchy(arrow.core.Either)
+ *
+ * {:.beginner}
+ * beginner
+ *
```
Comment (i-walker) [inline modules/core/arrow-core-data/src/main/kotlin/arrow/core/Either.kt]:
```suggestion
```

delete that line

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/Either.kt [hunk 14]
```diff
@@ -6,11 +6,620 @@ import arrow.core.Either.Right
...
+ * fun reciprocal(i: Int): Double =
+ *   if (i == 0) throw IllegalArgumentException("Cannot take reciprocal of 0.")
+ *   else 1.0 / i
+ *
+ * fun stringify(d: Double): String = d.toString()
+ * //sampleEnd
+ * ```
+ *
+ * Instead, let's make the fact that some of our functions can fail explicit in the return type.
+ *
```
Comment (i-walker) [inline modules/core/arrow-core-data/src/main/kotlin/arrow/core/Either.kt]: One section is missing here:
```suggestion
 * ```kotlin:ank
 * import arrow.core.Either
 * import arrow.core.flatMap
 * import arrow.core.left
 * import arrow.core.right
 *
 * //sampleStart
 * // Either Style
 * fun parse(s: String): Either<NumberFormatException, Int> =
 *   if (s.matches(Regex("-?[0-9]+"))) Either.Right(s.toInt())
 *   else Either.Left(NumberFormatException("$s is not a valid integer."))
 *
 * fun reciprocal(i: Int): Either<IllegalArgumentException, Double> =
 *   if (i == 0) Either.Left(IllegalArgumentException("Cannot take reciprocal of 0."))
 *   else Either.Right(1.0 / i)
 *
 * fun stringify(d: Double): String = d.toString()
 *
 * fun magic(s: String): Either<Exception, String> =
 *   parse(s).flatMap { reciprocal(it) }.map { stringify(it) }
 * //sampleEnd
 * ```
 * 
 * These calls to `parse` returns a `Left` and `Right` value
 *
```

**Comment by i-walker** [comment]
looks good. Now copy-paste the content of each data type to its ReadMe file in `arrow-docs`and your good to go :) !!!

**Comment by maruiz81** [comment]
all done!

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/Either.kt [hunk 15]
```diff
@@ -6,11 +6,630 @@ import arrow.core.Either.Right
...
+ * //sampleStart
+ * val value = null.rightIfNotNull { "left" }
+ * //sampleEnd
+ *
+ * fun main() {
+ *  println("value = $value")
+ * }
+ * ```
+ *
```
[the remark this answers is not in the document]
Comment (i-walker) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/core/Either.kt]: Not resolved IMO

[the remark this answers is not in the document]
Comment (i-walker) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/core/Either.kt]: not resolved IMO

File: modules/docs/arrow-docs/docs/docs/arrow/core/either/README.md [hunk 16]
```diff
@@ -1,19 +1,8 @@
...
-layout: docs
-title: Either
-permalink: /docs/arrow/core/either/
-redirect_from:
-  - /docs/datatypes/either/
-video: q6HpChSq-xc
----
-
-## Either
+ank_macro_hierarchy(arrow.core.Either)
```
Comment (i-walker) [inline modules/docs/arrow-docs/docs/docs/arrow/core/either/README.md]: `ank_macro_hierarchy` is only valid in the KDocs

File: modules/docs/arrow-docs/docs/docs/arrow/core/either/README.md [hunk 17]
```diff
@@ -1,19 +1,8 @@
----
```
Comment (i-walker) [inline modules/docs/arrow-docs/docs/docs/arrow/core/either/README.md]: The first block including the name of the data type must be in the README

File: modules/docs/arrow-docs/docs/docs/arrow/data/validated/README.md [hunk 18]
```diff
@@ -203,104 +320,272 @@ If you do want error accumulation but occasionally run into places where sequent
...
 }
+//sampleEnd
 ```
 
 `Rules` defines abstract behaviors that can be composed and have access to the scope of `ApplicativeError` where we can invoke `just` to lift values in to the positive result and `raiseError` into the error context.
 
 Once we have such abstract algebra defined we can simply materialize it to data types that support different error strategies:
 
-*Error accumulation*
+Error accumulation*
```
Comment (i-walker) [inline modules/docs/arrow-docs/docs/docs/arrow/data/validated/README.md]:
```suggestion
*Error accumulation*
```

File: modules/docs/arrow-docs/docs/docs/arrow/data/validated/README.md [hunk 19]
```diff
@@ -203,104 +320,272 @@ If you do want error accumulation but occasionally run into places where sequent
...
+     FormField("Valid Email Label", "getlost@nowhere.com")
+   ).map { it.validateEmail() }
+   }
+//sampleEnd
+fun main() {
+ println(value)
 }
 ```
-*Fail Fast*
+Fail Fast*
```
Comment (i-walker) [inline modules/docs/arrow-docs/docs/docs/arrow/data/validated/README.md]:
```suggestion
*Fail Fast*
```

File: modules/docs/arrow-docs/docs/docs/arrow/data/validated/README.md [hunk 20]
```diff
@@ -154,47 +161,166 @@ type alias `ValidatedNel<E, A>` is provided.
...
+     is Some -> s.t.valid()
+     is None -> ConfigError.ParseConfig(key).invalid()
+    }
+   is None -> Validated.Invalid(ConfigError.MissingConfig(key))
+  }
+ }
+}
+
+fun <E, A, B, C> parallelValidate(v1: Validated<E, A>, v2: Validated<E, B>, f: (A, B) -> C): Validated<E, C> {
+ return when {
```
Comment (i-walker) [inline modules/docs/arrow-docs/docs/docs/arrow/data/validated/README.md]: here, too.

File: modules/docs/arrow-docs/docs/docs/arrow/core/either/README.md [hunk 21]
```diff
@@ -309,12 +547,28 @@ be wrapped to `Either.Right`.
...
 Example:
 
-```kotlin:ank
-"value".rightIfNotNull { "left" }
+```kotlin:ank:playground
+import arrow.core.rightIfNotNull
+
+val value =
+//sampleStart
+ "value".rightIfNotNull { "left" }
```
Comment (i-walker) [inline modules/docs/arrow-docs/docs/docs/arrow/core/either/README.md]: There is something wrong here check your build

File: modules/docs/arrow-docs/docs/docs/arrow/core/either/README.md [hunk 22]
```diff
@@ -323,58 +577,95 @@ If the value is not null than it will be transformed to the specified `Either.Le
...
 
 ```kotlin:ank:replace
-import arrow.reflect.*
-import arrow.core.*
+import arrow.reflect.DataType
+import arrow.reflect.tcMarkdownList
+import arrow.core.Either
 
 DataType(Either::class).tcMarkdownList()
 ```
```
Comment (i-walker) [inline modules/docs/arrow-docs/docs/docs/arrow/core/either/README.md]: This guy was missing completely from our docs :) 
```suggestion
```

ank_macro_hierarchy(arrow.core.Either)
```

File: modules/docs/arrow-docs/docs/docs/arrow/core/either/README.md [hunk 23]
```diff
@@ -321,58 +577,94 @@ If the value is not null than it will be transformed to the specified `Either.Le
...
 
 ```kotlin:ank:replace
-import arrow.reflect.*
-import arrow.core.*
+import arrow.reflect.DataType
+import arrow.reflect.tcMarkdownList
+import arrow.core.Either
 
 DataType(Either::class).tcMarkdownList()
-```
```
Comment (i-walker) [inline modules/docs/arrow-docs/docs/docs/arrow/core/either/README.md]: why did you delete it?

**Comment by i-walker** [comment]
Feel free to merge, once it builds :)
